Better fee estimates

morcos opened the pull request:
This is a substantial improvement of fee estimation code built off #9942.
Although in many cases the estimates will not appear to differ by much from previous estimates, they will handle some situations much better.

A longer writeup of the original algorithm and the changes in this PR can be found here:
https://gist.github.com/morcos/d3637f015bc4e607e1fd10d8351e9f41

A summary of major changes:
- Estimates can now be given for up to a 1008 block target.
- Estimates are tracked on 3 different time horizons which allows for both the longer targets and lets the estimates adjust more quickly to changes in conditions.
- estimatefee is deprecated in favor of using only estimatesmartfee (already used by the GUI)
- estimateSmartFee now requires a 60% success rate at target/2, an 85% rate at target and a 95% rate at 2*target.
- estimates are by default conservative which requires the 95% rate for 2*target to be met at longer time horizons as well, but by choosing non-conservative estimates, the estimates are drastically reduced during periods of less transaction activity (such as the weekend).
- estimates are smarter about making sure enough data has been gathered in order to return a valid estimate (at least twice the number of blocks as the requested target much have been recorded)
- Fee rate buckets are half as big leading to a bit more precision.
- estimaterawfee is added so that customized logic can be implemented to analyze the raw data and calculate estimates.
- Transactions which leave the mempool due to eviction or other non-confirmed reasons now count as failures.  These are also saved at shutdown.

**A summary of open issues:**
- Estimates are not backwards compatible.  Currently when updating to this patch, any old fee estimates will be discarded and your node will have to be running for a while to provide new estimates.  It is a possibility to add logic to return the old estimates for targets 2-25 until the new estimates have gathered enough data.
- estimatefee is probably a roughly similar calculation to what it used to be, but I've made no effort to make sure the quality of the estimates it gives have not degraded, so infrastructure which doesn't support estimatesmartfee should update.
- The GUI needs to be updated to access the new range of estimates.  I would suggest providing options for targets of: 2,4,6,12,24,48,144,504,1008

morcos committed:
Expose estimaterawfee

Track information the ranges of fee rates that were used to calculate the fee estimates (the last range of fee rates in which the data points met the threshold and the first to fail) and provide an RPC call to return this information.

morcos committed:
minor refactor: explicitly track start of new bucket range and don't update curNearBucket on final loop.

morcos committed:
Change file format for fee estimates.

Move buckets and bucketMap to be stored as part of overall serialization of estimator.
Add some placeholder data so file format is only changed once.
Maintain 3 different TxConfirmStats with potential for different decays and scales.

morcos committed:
Refactor to update moving average on fly

morcos committed:
Make EstimateMedianVal smarter about small failures.

Instead of stopping if it encounters a "sufficient" number of transactions which don't meet the threshold for being confirmed within the target, it keeps looking to add more transactions to see if there is a temporary blip in the data.  This allows a smaller number of required data points.

morcos committed:
Rewrite estimateSmartFee

Change the logic of estimateSmartFee to check a 60% threshold at half the target, a 85% threshold at the target and a 95% threshold at double the target. Always check the shortest time horizon possible and ensure that estimates are monotonically decreasing.  Add a conservative mode, which makes sure that the 95% threshold is also met at longer time horizons as well.

morcos committed:
Track first recorded height

Track the first time we seen txs in a block that we have been tracking in our mempool. Used to evaluate validity of fee estimates for different targets.

morcos committed:
Change parameters for fee estimation and estimates on all 3 time horizons.

Make feerate buckets smaller (5% instead of 10%) and make the 3 different horizons have half lifes of 3 hours, 1 day and 1 week respectively.

morcos committed:
Track failures in fee estimation.

Start tracking transactions which fail to confirm within the target and are then evicted or otherwise leave mempool.

Fix slight error in unit test.

morcos committed:
Clean up fee estimate debug printing

morcos committed:
Historical block span

Store in fee estimate file the block span for which we were tracking estimates, so we know what targets we can successfully evaluate with the data in the file. When restarting use either this historical block span to set valid range of targets until our current span of tracking estimates is just as long.

morcos committed:
Introduce a scale factor

For the per confirmation number tracking of data, introduce a scale factor so that in the longer horizones confirmations are bucketed together at a resolution of the scale.  (instead of 1008 individual data points for each fee bucket, have 42 data points each covering 24 different confirmation values.. (1-24), (25-48), etc.. )

morcos committed:
minor cleanup: remove unnecessary variable

morcos committed:
Comments and improved documentation

jonasschnelli commented:
Great!
Concept ACK.

> The GUI needs to be updated to access the new range of estimates. I would suggest providing options for targets of: 2,4,6,12,24,48,144,504,1008

I'll have a look at this soon.

jonasschnelli commented:
ping @crwatkins (he presented some fee estimation overview at the wallet standards group meeting in Berlin this month).

jlopp commented on src/rpc/mining.cpp:
```diff
@@ -853,15 +855,98 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+            "\nWARNING: This interface is unstable and may disappear or change!\n"
+            "\nWARNING: This is an advanced API call that is tightly coupled to the specific\n"
+            "         implementation of fee estimation. The parameters it can be called with\n"
+            "         and the results it returns will change if the internal implementaion changes.\n"
```
Nit: implementation

jlopp commented:
I'm not sure if this is a problem specific to your code, though requiring a leading 0 seems like a bug:

```
bitcoin-cli estimaterawfee 10 .85 0
error: Error parsing JSON:.85
```

sipa commented:
@jlopp .85 is not valid JSON.

jlopp commented on src/rpc/mining.cpp:
```diff
@@ -853,15 +855,98 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+    int nBlocks = request.params[0].get_int();
+    double threshold = 0.95;
+    if (!request.params[1].isNull())
+        threshold = request.params[1].get_real();
```
Would prefer more explicit range bounding on the threshold with an "Invalid threshold" error response; it currently allows numbers < 0 and > 1 and just returns a -1 feerate.

morcos replied:
The reason I didn't bother with that is I felt estimaterawfee should be a very low level feature that doesn't need all the user friendly bells and whistles..   But it could certainly be cleaned up to add that either in this PR or a follow on..

morcos commented:
clean rebase after #9942 was merged

crwatkins commented:
Thanks @jonasschnelli.  In Berlin, I noted that at least 50% of the wallets listed on bitcoin.org use Core as a fee estimation source directly or through other servers based on Core and any statistical or algorithmic biases might compound over time because of the large number of clients using the single source.   I presented some anecdotal, non-scientific observations about fees and some reasons that they might be artificially high.  Some of my concerns about the current esitmatefee were

* Slow to respond (2.5 day half life)
* Too high confidence (95%)
* No long term estimates (only 25 blocks)

I believe that the changes in this PR mitigate some of those concerns, and thus I am in favor of these changes.  Nice work!  

I'm still a little concerned about so many wallets "competing" using the same identical algorithm.  @morcos do you envision some devs (wallets) would use estimaterawfee to tweak the algorithm for their users?

jlopp commented:
@crwatkins I can only speak for BitGo, but we run custom compiled versions of Core with our own half life and confidence values. We then use the output as a baseline for a more complex fee algorithm that we adjust upward and downward based upon other data sources.

morcos commented:
@crwatkins Yes that was the basic idea of including estimaterawfee (that and it's useful for debugging).  The exact strategy you use for determining what fee you want to put on a transaction depends on so many factors that its hard to have a one size fits all solution, but thats exactly what Bitcoin Core has to try to ship.  So rather than attempt to defend why the algorithm I picked is the one algorithm to rule them all, I thought it would also make sense to expose a way to examine all the collected data and let people build their own algorithms if they'd like to.

crwatkins commented:
@morcos Exclellent. Thanks!

Leviathn commented:
Reviewing

TheBlueMatt reviewed (commented):
Looks good. With the exception of the fallback during upgrade and the one strange sharp edge LGTM. Needs testing because the code being correct isnt really enough, but no doubt better than today's estimates.

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -828,16 +829,16 @@ UniValue estimatefee(const JSONRPCRequest& request)
             "\nArguments:\n"
-            "1. nblocks     (numeric)\n"
+            "1. nblocks       (numeric)\n"
+            "2. conservative  (bool, optional, default=true) Whether to return a more conservative estimate calculated from a longer history\n"
```
It's not clear to me as the API reader what "conservative" means here, depending on if I'm worrying about over-paying or waiting too long, or being conservative in the amount of data I'm using to estimate.

suggested: "Whether to return a possibly higher feerate estimate calculated from a longer history"

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -854,15 +855,98 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+            "2. threshold   (numeric, optional) The proportion of transactions in a given feerate range that must have been\n"
+            "               confirmed within nblocks in order to consider those feerates as high enough and proceed to check\n"
+            "               lower buckets.  Default: 0.95\n"
+            "3. horizon     (numeric, optional) How long a history of estimates to consider. 0=short, 1=medium, 2=long.\n"
```
perhaps mention how long each history is

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -854,15 +855,98 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+            "  \"decay\" : x.x,          (numeric) exponential decay for historical moving average of confirmation data\n"
+            "  \"scale\" : x,            (numeric) The resolution of confirmation targets at this time horizon\n"
+            "  \"pass.\"                 information about the last range of feerates to succeed in meeting the threshold\n"
+            "  \"fail.\"                 information about the first range of feerates to fail to meet the threshold\n"
```
nit: s/first/highest/

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -854,15 +855,98 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+            "  \"feerate\" : x.x,        (numeric) estimate fee-per-kilobyte (in BTC)\n"
+            "  \"decay\" : x.x,          (numeric) exponential decay for historical moving average of confirmation data\n"
+            "  \"scale\" : x,            (numeric) The resolution of confirmation targets at this time horizon\n"
+            "  \"pass.\"                 information about the last range of feerates to succeed in meeting the threshold\n"
```
nit: s/last/lowest/

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -854,15 +855,98 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+            "  \"startrange\" : x.x,     (numeric) start of feerate range\n"
+            "  \"endrange\" : x.x,       (numeric) end of feerate range\n"
+            "  \"withintarget\" : x.x,   (numeric) number of txs over history horizon in the feerate range that were confirmed within target\n"
+            "  \"totalconfirmed\" : x.x, (numeric) number of txs over history horizon in the feerate range total\n"
```
s/total/that were confirmed at any point/

instagibbs commented on test/functional/smartfees.py:
```diff
@@ -116,8 +116,8 @@ def check_estimates(node, fees_seen, max_invalid, print_estimates = True):
-            # estimatesmartfee should return the same result
-            assert_equal(node.estimatesmartfee(i+1)["feerate"], e)
+            if i >= 13:  # for n>=14 estimatesmartfee(n/2) should be at least as high as estimatefee(n)
+                assert(node.estimatesmartfee((i+1)//2)["feerate"] > float(e) - delta)
```
could also do a quick check that the transitive property is true for `conservative` estimates

instagibbs commented on src/policy/fees.h:
```diff
@@ -96,6 +100,55 @@ static const double FEE_SPACING = 1.1;
+
+    /** Require an avg of 0.1 tx in the combined feerate bucket per block to have stat significance */
+    static constexpr double SUFFICIENT_FEETXS = 0.1;
+    /** Require an avg of 0.5 tx when using short decay since there are less blocks considered*/
```
grammar nit: s/less/fewer/

instagibbs reviewed (approved):
utACK 57e71c819d7653555b048288eee717d53d493924 provided `conservative` is defined to the user

If I find the time I'll do some real-world testing once the buckets fill up.

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -854,15 +855,98 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() < 1|| request.params.size() > 3)
+        throw std::runtime_error(
+            "estimaterawfee nblocks\n"
```
missing `threshold`

TheBlueMatt commented on src/policy/fees.cpp:
```diff
@@ -503,67 +586,220 @@ void CBlockPolicyEstimator::processBlock(unsigned int nBlockHeight,
+            median = doubleEst;
+        }
+
+        if (conservative) {
```
I believe we may want to add a check that the previous code actually got a result here. And, generally, that all of the four estimate*Fee calls here actually succeed. It seems super strange to accept only one result as fine if the rest fail (eg on startup you may get a response from short horizon on SUFFICIENT_TXS_SHORT, but not on medium/long, also @sdaftuar saw a case this morning where he had an offline node for a while which succeeded in getting a result for conservative but not for non-conservative for  a 3 conftarget, I believe because estimateConservativeFee gave a result from the medium/long horizons when the short horizon could not).

morcos replied:
After offline discussion, it appears the bigger issue is what to do when the short time horizon estimates have decayed to not being able to give you an answer.  It seems like what happens in conservative mode is probably the right thing where you take advantage of the fact that the longer time horizon estimates have not decayed yet.  We probably want the ability to utilize this as a fall back in non-conservative mode as well.  This can happen either from shutting down a node for a couple of days which means the short time horizon will decay too much, or even just from losing network activity from something like a laptop over the weekend.
As it is, I think its probably still a strict improvement to existing estimates, but could use further refinement.

morcos commented on src/policy/fees.cpp:
```diff
@@ -503,67 +586,220 @@ void CBlockPolicyEstimator::processBlock(unsigned int nBlockHeight,
+                if (!conservative) {
+                    // When we aren't using conservative estimates, if a lower confTarget from a more recent horizon returns a lower answer use it.
+                    double medMax = feeStats->EstimateMedianVal(feeStats->GetMaxConfirms(), SUFFICIENT_FEETXS, successThreshold, true, nBestSeenHeight);
+                    if (estimate == -1 ||  medMax < estimate) {
```
in all 3 of these replacement evaluations (also line 709 and line 722), we probably don't want to do the replacement if the new estimate is -1, should be rare.

morcos commented on src/policy/fees.cpp:
```diff
@@ -503,67 +586,220 @@ void CBlockPolicyEstimator::processBlock(unsigned int nBlockHeight,
+            return CFeeRate(0);
+
+        if ((unsigned int)confTarget > maxUsableEstimate) {
+            confTarget = maxUsableEstimate;
```
The design of estimateSmartFee is to be used by the wallet to intelligently as possible automatically put a fee on a transaction being sent.  In the GUI, the slider clearly indicates what target the fee is being returned for.  For the RPC call estimatesmartfee, this is also clearly returned.  The difficulty comes with automatic estimation that happens from a sendtoaddress or sendmany RPC call.  If in the future we add a way to select a transaction confirm target with the desired RPC call, perhaps it would make sense to fail, but for now it seems to me that its more important for the transaction to go out, even if its paying a fee for a much faster target.

morcos commented on src/policy/fees.cpp:
```diff
@@ -576,14 +812,24 @@ CFeeRate CBlockPolicyEstimator::estimateSmartFee(int confTarget, int *answerFoun
+        fileout << 149900; // version required to read: 0.14.99 or later
         fileout << CLIENT_VERSION; // version that wrote the file
         fileout << nBestSeenHeight;
+        if (BlockSpan() > HistoricalBlockSpan()/2) {
```
After discussion, I agree that this logic could be more intelligent, but I think that can be left for a later improvement.

morcos commented on src/policy/fees.cpp:
```diff
@@ -596,17 +608,79 @@ bool CBlockPolicyEstimator::Read(CAutoFile& filein)
+            std::map<double, unsigned int> tempMap;
+
+            std::unique_ptr<TxConfirmStats> tempFeeStats(new TxConfirmStats(tempBuckets, tempMap, MAX_BLOCK_CONFIRMS, tempDecay));
+            tempFeeStats->Read(filein, nVersionThatWrote, tempNum);
```
leaving for a later improvement

morcos commented on src/txmempool.cpp:
```diff
@@ -448,7 +448,7 @@ void CTxMemPool::removeUnchecked(txiter it, MemPoolRemovalReason reason)
     mapTx.erase(it);
     nTransactionsUpdated++;
-    if (minerPolicyEstimator) {minerPolicyEstimator->removeTx(hash);}
+    if (minerPolicyEstimator) {minerPolicyEstimator->removeTx(hash, false);}
```
This will eventually be cleaned up when fee estimation listens to CValidationInterface

morcos commented:
Addressed the comments in fixup commmits e95eca7 ([smarterfee.ver5](https://github.com/morcos/bitcoin/commits/smarterfee.ver5)) and then squashed --> 9a34419 ([smarterfee.ver5.squash](https://github.com/morcos/bitcoin/commits/smarterfee.ver5.squash))

sipa commented on src/policy/fees.h:
```diff
@@ -159,6 +171,10 @@ class CBlockPolicyEstimator
 {
+private:
+    static constexpr double MAX_FILTER_FEERATE = 1e7;
+    static constexpr double FEE_FILTER_SPACING = 1.1;
```
Can you add a comment to explain the relation with the CBlockPolicyEstimator::FEE_SPACE constant?

morcos replied:
yeah, there is no relation.
when I made the filter spacing, I just copied the other one arbitrarily, but then I decided if I was changing the estimate spacing, there was no reason to also make a change to the filter spacing.   I can add a comment.

sipa commented on src/rpc/mining.cpp:
```diff
@@ -854,15 +858,98 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+            "\nWARNING: This interface is unstable and may disappear or change!\n"
+            "\nWARNING: This is an advanced API call that is tightly coupled to the specific\n"
+            "         implementation of fee estimation. The parameters it can be called with\n"
+            "         and the results it returns will change if the internal implementaion changes.\n"
```
@jlopp's comment hasn't been addressed yet: "nit: Implementation".

morcos replied:
oops

sipa commented on src/rpc/mining.cpp:
```diff
@@ -863,6 +863,79 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+        if (horizonInt < 0 || horizonInt > 2) {
+            throw JSONRPCError(RPC_TYPE_ERROR, "Invalid horizon for fee estimates");
+        }
+        else {
```
Style nit: `else` on the same line as `}`

morcos replied:
kk

sipa commented on src/rpc/mining.cpp:
```diff
@@ -863,6 +863,79 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+            "2. threshold   (numeric, optional) The proportion of transactions in a given feerate range that must have been\n"
+            "               confirmed within nblocks in order to consider those feerates as high enough and proceed to check\n"
+            "               lower buckets.  Default: 0.95\n"
+            "3. horizon     (numeric, optional) How long a history of estimates to consider. 0=short, 1=medium, 2=long.\n"
```
To make the RPC interface less dependent on the specific implementation, perhaps don't make horizon a parameter, and just return information for all applicable horizons for the given value of nblocks?

morcos replied:
slightly prefer existing, but I could change it if prevailing opinion is otherwise..
estimaterawfee is meant to be implementation dependent

sipa commented on src/policy/fees.cpp:
```diff
@@ -658,31 +658,97 @@ CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThr
+
+double CBlockPolicyEstimator::estimateCombinedFee(unsigned int confTarget, double successThreshold, bool conservative) const
+{
+    double estimate = -1;
```
Can you write this function as a loop over the 3 horizons, instead of a cascaded if/then/else?

morcos replied:
I can take a pass at rewriting the logic to be a bit clearer, but I'm not sure a loop is the right thing.  There are basically 6 different combinations of estimates you might want to check depending on what range your target is in and whether or not you want to check shorter horizons.  But the logic might be cleaner if those are just spelled out.

sipa commented on src/policy/fees.h:
```diff
@@ -173,6 +173,10 @@ class CBlockPolicyEstimator
     CFeeRate minTrackedFee;    //!< Passed to constructor to avoid dependency on main
     unsigned int nBestSeenHeight;
+    unsigned int firstRecordedHeight;
+    unsigned int historicalFirst;
```
Nit: unused until 2 commits later.

morcos replied:
I wanted to get the file format change over and done with...

sipa commented on src/policy/fees.cpp:
```diff
@@ -663,6 +664,29 @@ CFeeRate CBlockPolicyEstimator::estimateRawFee(int confTarget, double successThr
+unsigned int CBlockPolicyEstimator::MaxUsableEstimate() const
+{
+    // Block spans are divided by 2 to make sure there are enough potential failing data points for the estimate
+    return std::min(longStats->GetMaxConfirms(), std::max(BlockSpan(), HistoricalBlockSpan()) / 2);
```
Nit: If you'd return which of the two BlockSpans was chosen in a pair or a bool, you wouldn't need to repeat the decision logic in the debug print statement of processBlock.

morcos replied:
ok

morcos replied:
after actually doing this, it seemed uglier, so leaving it as is

sipa reviewed (commented):
Code review ACK, with some nits. I haven't reviewed the changes to the estimation algorithm itself.

morcos commented on src/policy/fees.cpp:
```diff
@@ -580,10 +587,15 @@ bool CBlockPolicyEstimator::Write(CAutoFile& fileout) const
     try {
         LOCK(cs_feeEstimator);
-        fileout << 139900; // version required to read: 0.13.99 or later
+        fileout << 149900; // version required to read: 0.14.99 or later
```
no reason..., but this also seems simple enough?

morcos commented on src/policy/fees.h:
```diff
@@ -141,19 +207,38 @@ class CBlockPolicyEstimator
 
     /** Classes to track historical data on transaction confirmations */
     TxConfirmStats* feeStats;
+    TxConfirmStats* shortStats;
```
This was the change made in #9942, and I couldn't get unique_ptrs to work while also moving the code to the cpp file.  I now also have other ideas how to clean up the wonky design of TxConfirmStats inside CBlockPolicyEstimator, but I don't want to do it now and churn this PR.

morcos commented:
OK.  I think I addressed remaining comments.
I also tried to clean up the logic of estimateCombinedFee to be a bit clearer (although still not a loop)
I also changed the logic to be smarter about returning an answer if one can be found on any horizon, instead of returning -1.   
Fixup commits are squashed.

abe88e4 ([smarterfee.ver7](https://github.com/morcos/bitcoin/commits/smarterfee.ver7))  --> 2d2e170 ([smarterfee.ver7.squash](https://github.com/morcos/bitcoin/commits/smarterfee.ver7.squash))

instagibbs commented:
re-utACK https://github.com/bitcoin/bitcoin/pull/10199/commits/2d2e17052c389948c511d2b5d41c4cc243cdc145

sipa reviewed (commented):
utACK

gmaxwell reviewed (approved):
ACK.

morcos commented on src/rpc/mining.cpp:
```diff
@@ -863,6 +863,78 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
+
+    result.push_back(Pair("feerate", feeRate == CFeeRate(0) ? -1.0 : ValueFromAmount(feeRate.GetFeePerK())));
+    result.push_back(Pair("decay", buckets.decay));
+    result.push_back(Pair("pass.startrange", round(buckets.pass.start)));
```
ah, that makes sense.  i will make this change

morcos committed:
Make more json-like output from estimaterawfee

morcos commented:
Added an extra commit to improve JSON output as per @sipa.  Was kind of annoying to squash, so will leave it if that's ok?

@laanwj I think this is reviewed enough that we'd be better off merging and getting some more experimentation with the estimates..    It takes 2 full weeks for the longest estimates to even be exposed.  

The biggest open questions are whether I should add an ability to use old estimates during transition and whether we are going to build in GUI support.

The tests could also use some improvement but I don't think they're notably worse than they were.

sipa commented:
utACK 38bc1ec4a473b5bd9b7bbce7b20a11e8edfe4b4c

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/10199/commits/38bc1ec4a473b5bd9b7bbce7b20a11e8edfe4b4c

RHavar commented:
I tried reading the description from help, and I don't really understand this:

> ...and return the number of blocks for which the estimate is valid.

but from the actual return:

>   "blocks" : n         (numeric) block number where estimate was found


Without reading the code, it's not clear what `blocks` means

instagibbs commented:
@RHavar without actually looking at the code, I believe this refers to the "cheapest bucket" block which that feerate was found.

RHavar commented:
@instagibbs  Yeah, you seem right. But I don't think that was inferable from the help information :P

What's the motivation for `estimatefee` returning a different value than `estimatesmartfee` ? 

`$ bitcoin-cli estimatefee 6
0.00254275`


`bitcoin-cli estimatesmartfee 6
{
  "feerate": 0.00208824,
  "blocks": 6
}`


?

RHavar commented:
I've updated https://estimatefee.com/  to now use `estimatesmartfee $n` so if you're interested you can take a look there.

The bitcoin node has only been up a few hours, so it'll probably take a while to collect enough information.

P.S. you kinda broke my domain name lol

morcos commented:
@Rhavar you might be right that the description is not the clearest.
blocks refers to the target for which you are getting an estimate, which is an expected number of blocks within which you will be confirmed.  Occasionally you ask for an estimate at one target but it returns you an estimate for a different target (b/c it is unable to give you an estimate for the target you asked for)

RHavar commented:
@morcos  In that case I'd suggest renaming it from "blocks" to "target" and changing the description to what you wrote above :P

runn1ng commented:
Not sure where to put it, but just FYI

I made a simple tool that looks up smart fees once per 10 minutes and puts it here

https://estimatesmartfee.com/html.html

or

https://estimatesmartfee.com/json.json

If you want to use it for testing or something

jb55 commented:
also FYIs, I put together a small bash script that uses this for quickly grabbing fees from the command line: https://bitcoin.stackexchange.com/a/61348/1235 

thanks!
